File: java/app/src/main/java/com/technogise/foundation/geometry/Line.java
package com.technogise.foundation.geometry;

public record Line(Point a, Point b) {
    public double run() {
        return b.x() - a.x();
    }

    public double rise() {
        return b.y() - a.y();
    }

    public double length() {
        double dx = run();
        double dy = rise();

        return Math.sqrt(dx*dx + dy*dy);
    }

    public double slope() {
        return rise() / run();
    }

    public boolean isParallelTo(Line other) {
        return this.slope() == other.slope();
    }

    public boolean isPerpendicularTo(Line other) {
        return this.slope() * other.slope() == -1;
    }
}

File: core/go/geometry/line.go
package geometry

import "math"

type Line struct {
	Start Point
	End   Point
}

func (l Line) Rise() float64 {
	return l.End.Y - l.Start.Y
}

func (l Line) Run() float64 {
	return l.End.X - l.Start.X
}

func (l Line) Length() float64 {
	return math.Sqrt(l.Rise()*l.Rise() + l.Run()*l.Run())
}

func (l Line) Slope() float64 {
	return l.Rise() / l.Run()
}

func (l Line) IsParallelTo(other Line) bool {
	return l.Slope() == other.Slope()
}

func (l Line) IsPerpendicularTo(other Line) bool {
	return l.Slope()*other.Slope() == -1
}
